src/network/clientpackethandler.cpp

```diff
@@ -1485,46 +1490,71 @@ void Client::handleCommand_PlayerSpeed(NetworkPacket *pkt)
 void Client::handleCommand_MediaPush(NetworkPacket *pkt)
 {
 	std::string raw_hash, filename, filedata;
+	u32 token;
 	bool cached;
 
-	*pkt >> raw_hash >> filename >> cached;
-	filedata = pkt->readLongString();
+	if (m_proto_ver >= 40) {
+		*pkt >> raw_hash >> filename >> token >> cached;
+	} else {
+		*pkt >> raw_hash >> filename >> cached;
+		filedata = pkt->readLongString();
+	}
 
-	if (raw_hash.size() != 20 || filedata.empty() || filename.empty() ||
+	bool data_empty = m_proto_ver < 40 && filedata.empty();
+	if (raw_hash.size() != 20 || data_empty || filename.empty() ||
 			!string_allowed(filename, TEXTURENAME_ALLOWED_CHARS)) {
 		throw PacketError("Illegal filename, data or hash");
 	}
 
-	verbosestream << "Server pushes media file \"" << filename << "\" with "
-		<< filedata.size() << " bytes of data (cached=" << cached
-		<< ")" << std::endl;
+	verbosestream << "Server pushes media file \"" << filename << "\" ";
+	if (filedata.empty())
+		verbosestream << "to be fetched ";
+	else
+		verbosestream << "with " << filedata.size() << " bytes ";
+	verbosestream << "(cached=" << cached << ")" << std::endl;
 
 	if (m_media_pushed_files.count(filename) != 0) {
-		// Silently ignore for synchronization purposes
+		// Silently ignore. Previously this was for sync purposes, but even in
+		// new versions media cannot be replaced at runtime.
 		return;
 	}
 
-	// Compute and check checksum of data
-	std::string computed_hash;
-	{
-		SHA1 ctx;
-		ctx.addBytes(filedata.c_str(), filedata.size());
-		unsigned char *buf = ctx.getDigest();
-		computed_hash.assign((char*) buf, 20);
-		free(buf);
-	}
-	if (raw_hash != computed_hash) {
-		verbosestream << "Hash of file data mismatches, ignoring." << std::endl;
+	if (!filedata.empty()) {
+		// LEGACY CODEPATH
+		// Compute and check checksum of data
+		std::string computed_hash;
+		{
+			SHA1 ctx;
+			ctx.addBytes(filedata.c_str(), filedata.size());
+			unsigned char *buf = ctx.getDigest();
+			computed_hash.assign((char*) buf, 20);
+			free(buf);
+		}
+		if (raw_hash != computed_hash) {
+			verbosestream << "Hash of file data mismatches, ignoring." << std::endl;
+			return;
+		}
+
+		// Actually load media
+		loadMedia(filedata, filename, true);
+		m_media_pushed_files.insert(filename);
+
+		// Cache file for the next time when this client joins the same server
+		if (cached)
+			clientMediaUpdateCache(raw_hash, filedata);
 		return;
 	}
 
-	// Actually load media
-	loadMedia(filedata, filename, true);
 	m_media_pushed_files.insert(filename);
 
-	// Cache file for the next time when this client joins the same server
-	if (cached)
-		clientMediaUpdateCache(raw_hash, filedata);
+	// create a downloader for this file
+	auto downloader = new SingleMediaDownloader(cached);
+	m_pending_media_downloads[token] = downloader;
+	downloader->addFile(filename, raw_hash);
+	for (const auto &baseurl : m_remote_media_servers)
+		downloader->addRemoteServer(baseurl);
```

Comment:
Do you really need to copy the `m_remote_media_servers` vector? Have you thought about using a `shared_ptr`, or maybe just a raw pointer to `m_remote_media_servers`?
If copying is really needed, at least just set all remote servers at once to avoid remallocs of the vector.

Reply:
I do not think shaving off a copy of zero to one strings sized less than 64 bytes is worth it.

Reply:
You're right.
I forgot how few remote servers there usually are.